### ISSUE TITLE
custom string types

### ISSUE THREAD

**Issue by psifertex** [Vector35/binaryninja-api#1334]
Long-term we'd like to allow for custom string formats. This would include:

- [ ] custom recognizers for string search
- [ ] reprs for annotation display
- [ ] potentially allow modifications? 

Not on the immediate-term todo list though.

**Comment by xusheng6** [comment]
Related to: https://github.com/Vector35/binaryninja-api/issues/2222

**Comment by utkonos** [comment]
@psifertex Yes, I think you're right that https://github.com/Vector35/binaryninja-api/issues/5794 is a component of this, if I am reading it correctly. The need I have is, for example, is when an encrypted string is located in data. That data symbol is type `char*` as a simple example. That data symbol then appears anywhere it gets passed around in the code, in functions. At the "source" and in all the binary ninja annotations `{"foo"}` in the code, the string is currently displayed as the encrypted string. My first goal is to write some Python that decrypts all the strings in the binary. Today I would put the output from that script, the plaintext string, in a comment in places where it appears. Sometimes this becomes tedious because a string is used in more than one place. If the string itself, in binary ninja, had some kind of display layer where I could put the decrypted representation, and the decrypted representation would appear anywhere binary ninja adds it's own `{"foo"}` annotations, that would be the minimum. It would then be a "nice to have" to be able to differentiate these custom display strings by a different color of text. And a "really nice to have" would be a custom color field so that I can have different types of strings and make it obvious what they are visually.

**Comment by utkonos** [comment]
One more thing: the minimum needs to be read and written via API so that it can be connected to a Python plugin's output.

**Comment by utkonos** [comment]
Sorry, one more thought: if you do implement a custom color feature for string display, could you have it encompass comments as well? I have this other issue about that idea:
https://github.com/Vector35/binaryninja-api/issues/5251

[#AngryFruitSalad](https://zvon.org/comp/r/ref-Jargon_file.html#Terms~angry_fruit_salad) 😆

**Comment by psifertex** [comment]
I don't think we would implement custom color feature for string displays, but the new Render Layer APIs enable you to do this yourself as I understand it right now. In fact, it's even possible you could implement a form of this string deobfuscation using that feature as a display-only level UI feature now.

https://github.com/Vector35/binaryninja-api/blob/dev/python/examples/args_render_layer.py
https://github.com/Vector35/binaryninja-api/blob/dev/python/examples/follow_reg_render_layer.py

**Comment by utkonos** [comment]
Thanks. Will check that out. Colors are a nice to have for the deobfuscated/decrypted strings. But just having the plaintext follow the actual string around in the various locations like Variable pane, annotations, etc etc is the main goal.

**Comment by psifertex** [comment]
Right, for that you'd need the full implementation but being able to get it viewable in linear decompilation is already a big improvement vs just adding a comment

**Comment by psifertex** [comment]
Added in builds >= 5.2.8461.

Remaining feature of being able to search for custom string encodings split into a new issue #7540

**Comment by mon** [comment]
How do you use the custom strings? The documentation still says
> There is currently no support for custom types with different encoding mechanisms. Follow #1334 for more updates on when that feature is added

Trying to display Japanese strings in a binary encoded in cp932.

**Comment by plafosse** [comment]
Oh thanks for pointing that out. We need to get our documentation updated. Please see our blog post! https://binary.ninja/2025/11/13/binary-ninja-5.2-io.html#custom-strings--constants
